apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Span.java

```diff
@@ -413,10 +413,11 @@ private boolean tryToCompressRegular(Span sibling) {
     }
 
     private boolean tryToCompressComposite(Span sibling) {
-        String compressionStrategy;
-        do {
-            compressionStrategy = composite.getCompressionStrategy();
-        } while (compressionStrategy == null);
+        String compressionStrategy = composite.getCompressionStrategy();
+        if (compressionStrategy == null) {
+            //lose the compression rather than retry, so that the application proceeds with minimum delay
+            return false;
+        }
```

Comment:
Thank you for working on this! I'd be happy to test the workaround.

I was wondering if the retry could have a kind of fail-safe, aborting after a fixed number of retries. Something similar to:

```suggestion
        String compressionStrategy;
        for (int i = 0; i < 100 && compressionStrategy == null; ++i) {
            compressionStrategy = composite.getCompressionStrategy();
        }
        if (compressionStrategy == null) {
            // lose the compression rather than retry, so that the application proceeds with minimum delay
          return false;
        }
```

But I assume the workaround will be removed soon, once the root cause is found?

Reply:
Let's see how it looks when you try it. We might prefer to leave the fast fail so that there is minimum impact to the application. When we've eliminated premature recycling, this concurrent race should happen rarely. At most I think I'd retry only once